src/Proto3/Wire/Encode.hs

```diff
@@ -211,61 +268,70 @@ double num d = fieldHeader num Fixed64 <> BB.doubleLE d
 -- >
 -- > fieldNumber 1 `enum` True <>
 -- > fieldNumber 2 `enum` Circle
-enum :: Enum e => FieldNumber -> e -> BB.Builder
+enum :: Enum e => FieldNumber -> e -> Builder
 enum num e = fieldHeader num Varint <> base128Varint (fromIntegral (fromEnum e))
 
 -- | Encode a UTF-8 string.
 --
 -- For example:
 --
 -- > fieldNumber 1 `string` "testing"
-string :: FieldNumber -> String -> BB.Builder
-string num = embedded num . BB.stringUtf8
+string :: FieldNumber -> String -> Builder
+string num = embedded num . stringUtf8
 
 -- | Encode lazy `Text` as UTF-8
 --
 -- For example:
 --
 -- > fieldNumber 1 `text` "testing"
-text :: FieldNumber -> Text.Lazy.Text -> BB.Builder
-text num txt = embedded num
-                        (BB.lazyByteString (Text.Lazy.Encoding.encodeUtf8 txt))
+text :: FieldNumber -> Text.Lazy.Text -> Builder
+text num txt =
+    embedded num (Builder (Sum len, Text.Lazy.Encoding.encodeUtf8Builder txt))
+  where
+    -- It would be nice to avoid actually allocating encoded chunks,
```

Comment:
Maybe we could provide a variant on this function which took the length as an additional argument. That way, if you do know it ahead of time for some reason, there's no need to recompute it from the chunks. What do you think?